src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -524,10 +528,18 @@ private ValueTask<FlushResult> WriteDataAndTrailersAsync(Http2Stream stream, in
 
             try
             {
-                _headersEnumerator.Initialize(headers);
                 _outgoingFrame.PrepareHeaders(Http2HeadersFrameFlags.END_STREAM, streamId);
-                var buffer = _headerEncodingBuffer.AsSpan();
-                var done = HPackHeaderWriter.BeginEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out var payloadLength);
+                HeaderWriteResult done = HeaderWriteResult.MoreHeaders;
+                int payloadLength;
+                do
+                {
+                    _headersEnumerator.Initialize(headers);
+                    _headerEncodingBuffer.ResetWrittenCount();
+                    var bufferSize = done == HeaderWriteResult.BufferTooSmall ? _headerEncodingBuffer.Capacity * 2 : _headerEncodingBuffer.Capacity;
+                    var buffer = _headerEncodingBuffer.GetSpan(bufferSize)[0..bufferSize];
```

Comment:
do we need this slice (`[0..bufferSize]`)? if we get more, is there any harm letting `HPackHeaderWriter` have it?

Reply:
Yes, otherwise it is not explicit if headers serialized after a large one will be in the same frame or separated, it all depends on if the header fits in the span returned by the `GetSpan` method csll

Reply:
probably warrants a comment to that effect, for the next reader